cloud/scope/powervs_machine_test.go

```diff
@@ -111,83 +122,1003 @@ func newPowerVSInstance(name, networkID, mac string) *models.PVMInstance {
+		g := NewWithT(t)
+		t.Run(tc.name, func(_ *testing.T) {
+			bucketName := tc.machineScope.bucketName()
+			g.Expect(bucketName).To(Equal(tc.expectedBucketName))
+		})
+	}
+}
+
+func TestBucketRegion(t *testing.T) {
+	testcases := []struct {
+		name               string
+		expectedportRegion string
+		machineScope       PowerVSMachineScope
+	}{
+		{
+			name:               "test - get region from cos instance",
+			expectedportRegion: region,
+			machineScope: PowerVSMachineScope{
+				IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+					Spec: infrav1beta2.IBMPowerVSClusterSpec{
+						CosInstance: &infrav1beta2.CosInstance{
+							BucketRegion: region,
+						},
+					},
+				},
+			},
+		}, {
+			name:               "test - get region from vpc",
+			expectedportRegion: region,
+			machineScope: PowerVSMachineScope{
+				IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+					Spec: infrav1beta2.IBMPowerVSClusterSpec{
+						VPC: &infrav1beta2.VPCResourceReference{
+							Region: ptr.To(region),
+						},
+					},
+				},
+			},
+		}, {
+			name:               "test - empty region (both cos instance and vpc source empty)",
+			expectedportRegion: "",
+			machineScope: PowerVSMachineScope{
+				IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+					Spec: infrav1beta2.IBMPowerVSClusterSpec{},
+				},
+			},
+		},
+	}
+
+	for _, tc := range testcases {
+		g := NewWithT(t)
+		t.Run(tc.name, func(_ *testing.T) {
+			region := tc.machineScope.bucketRegion()
+			g.Expect(region).To(Equal(tc.expectedportRegion))
+		})
+	}
+}
+
+func TestNewPowerVSMachineScope(t *testing.T) {
+	testCases := []struct {
+		name   string
+		params PowerVSMachineScopeParams
+	}{
+		{
+			name: "Error when Client in nil",
+			params: PowerVSMachineScopeParams{
+				Client: nil,
+			},
+		},
+		{
+			name: "Error when Machine in nil",
+			params: PowerVSMachineScopeParams{
+				Client:  testEnv.Client,
+				Machine: nil,
+			},
+		},
+		{
+			name: "Error when Cluster is nil",
+			params: PowerVSMachineScopeParams{
+				Client:  testEnv.Client,
+				Machine: newMachine(machineName),
+				Cluster: nil,
+			},
+		},
+		{
+			name: "Error when IBMPowerVSMachine is nil",
+			params: PowerVSMachineScopeParams{
+				Client:            testEnv.Client,
+				Machine:           newMachine(machineName),
+				Cluster:           newCluster(clusterName),
+				IBMPowerVSMachine: nil,
+			},
+		},
+		{
+			name: "Failed to get authenticator",
+			params: PowerVSMachineScopeParams{
+				Client:            testEnv.Client,
+				Machine:           newMachine(machineName),
+				Cluster:           newCluster(clusterName),
+				IBMPowerVSMachine: newPowerVSMachine(clusterName, machineName, core.StringPtr(pvsImage), core.StringPtr(pvsNetwork), true),
+			},
+		},
+	}
+	for _, tc := range testCases {
+		g := NewWithT(t)
+		t.Run(tc.name, func(_ *testing.T) {
+			_, err := NewPowerVSMachineScope(tc.params)
+			// Note: only error/failure cases covered
+			// TO-DO: cover success cases
```

Comment:
Can we cover success cases or do you think it required some code refactoring.

Reply:
Yes, it needs refactoring, It fails at this point in the code - https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/main/cloud/scope/powervs_machine.go#L183